model/notification.go

```diff
@@ -1,12 +1,146 @@
 package model
 
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+)
+
 type Notification struct {
-	UserID         int    `db:"user_id"`
-	AnnouncementID int    `db:"announcement_id"`
-	Status         string `db:"status"`
-	Receipt        string `db:"receipt"`
-	ReceiptStatus  string `db:"receipt_status"`
-	Timestamp      string `db:"timestamp"`
+	// From the notifications table.
+	ID              string `db:"id"`
+	UserID          int    `db:"user_id"`
+	AnnouncementID  int    `db:"announcement_id"`
+	Status          string `db:"status"`
+	LeaseExpiration int64  `db:"expiration_time"`
+	Receipt         string `db:"receipt"`
+	// From joined tables.
+	ExpoPushToken string `db:"expo_push_token"`
+	Title         string `db:"title"`
+	Body          string `db:"body"`
+}
+
+func EnqueueAnnouncementNotifications(db *sqlx.DB) {
+	log.Println("Starting to enqueue announcement notifications.")
+
+	err := transact(db, func(tx *sqlx.Tx) error {
+		insertQuery := `
+INSERT IGNORE into notifications (user_id, announcement_id, status)
+	SELECT users.id as user_id, announcements.id as announcement_id, "Queued" as status
+	FROM announcements
+	JOIN users ON users.conference_id = announcements.conference_id
+	WHERE sent = 0 AND send_time <= NOW() AND expo_push_token like "ExponentPushToken[%]"
+	ORDER BY send_time asc
+`
+		results := tx.MustExec(insertQuery)
+		notificationRows, err := results.RowsAffected()
+		if err != nil {
+			return fmt.Errorf("failed to insert into notifications table: %w", err)
+		}
+
+		updateQuery := `
+UPDATE announcements
+SET sent = 1
+WHERE id in (SELECT DISTINCT announcement_id FROM notifications) AND sent = 0
+`
+		results = tx.MustExec(updateQuery)
+		announcementRows, err := results.RowsAffected()
+		if err != nil {
+			return fmt.Errorf("failed to update announcements table: %w", err)
+		}
+
+		log.Printf("Enqueued %d notifications for %d announcements.\n", notificationRows, announcementRows)
+
+		return nil
+	})
+
+	if err != nil {
+		log.Println("Failed to enqueue announcement notifications.")
+		return
+	}
+
+}
+
+func SelectNotificationsToSend(db *sqlx.DB) ([]Notification, error) {
+	var notifications []Notification
+
+	currentTime := time.Now().Unix()
+	fiveMinFromNow := time.Now().Add(5 * time.Minute).Unix()
+
+	err := transact(db, func(tx *sqlx.Tx) error {
```

Comment:
Aside: Does this still need to be two separate queries in a transaction? Can it be a single update/select query?

Reply:
I thought about that for a bit when designing this, but I couldn't come up with a way to do it in one statement in MySQL. MySQL doesn't have a RETURNING clause for UPDATE statements like Postgres does.